### ISSUE TITLE
Add recursion-scheme-friendly ABTs

### ISSUE THREAD

**Issue by sellout** [slamdata/matryoshka#48]
Without all the associated machinery, we want something like this (but less hand-wavy):
```scala
sealed abstract class ABT[F[_], A]
final case class Var[F[_], A](v: String) extends ABT[F, A]
final case class Abs[F[_], A](v: String, term: A) extends ABT[F, A]

type BoundExpr[F[_], A] = HCoproduct(ABT, Expr, F, A)
```
which relies on us having mutual recursion, but gives us multi-sorted ABTs pretty simply, I think.

**Comment by jdegoes** [comment]
And maybe with a module-level existential encoding of `v` so users cannot synthesize their own variable identifiers out of thin air.

**Comment by sellout** [comment]
Yeah, I actually had a different encoding of vars, but thought it cluttered up the hand-waving.
